### DISCUSSION TITLE
Manifest with custom triplets example?

### DISCUSSION THREAD

**Discussion by jdx-john** [microsoft/vcpkg#19405] [Q&A]
We use custom triplets in our project, so currently we might run from the command-line:

```
vcpkg install tinyxml:x86-windows-custom install tinyxml:x64-windows-custom
vcpkg install gtest:x86-windows-custom
```

I was reading up on manifests, which I understand is still experimental: https://vcpkg.readthedocs.io/en/latest/users/manifests/

Is it possible to specify these exact combinations directly in my JSON file, if so what would that look like? Docs show examples of using the `platform` field:
`"platform": "!(windows & static) & !osx"`

But this doesn't seem a direct mapping to explicit triplet specification per-dependency.

May I ask, are substantial changes to manifests expected, is there a timeline when it is likely to drop experimental status and the command-line options change?

**Comment by autoantwort** [comment]
They are not experimental anymore. No it is not possible to specify the triplet inside the vcpkg.json file. To set the triplet do `vcpkg install --triplet=x86-windows-custom`

**Comment by jdx-john** [comment]
I see, I think... so I would still list triplets on the command-line as normal but it will gather which ports to install against those triplets from the manifest file?

In my example above, we don't build every port with every triplet, specifically we don't use 64-bit builds of several ports. I suppose building them doesn't matter too much, but can I control this in the manifest rather than do multiple install runs on each triplet?

Will the documentation be updated to reflect this is now a stable feature? I need to update my vcpkg installation clearly!

**Comment by jdx-john** [comment]
When I run `vcpkg help install` on the current version, it doesn't offer any note on `--x-manifest-root` either and https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md#command-line-interface still lists it as experimental.

Will the flag change - is `--x` reserved for experimental or does it just mean 'extra'?

**Comment by JackBoosY** [comment]
According to the [document](https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md#platform-field), the `platform` field only declared the supported platform, not the target triplet.
If you need to use the custom triplet:
1. with command line, please add `--triplet=YOUR_CUSTOM_TRIPLET`. See [document](https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md#--triplettriplet).
2. with cmake integration, please set `VCPKG_TARGET_TRIPET` to `YOUR_CUSTOM_TRIPLET`. See [document](https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md#vcpkg_target_triplet).
3. with msbuild integration, please add `YOUR_CUSTOM_TRIPLET` to the option `VcpkgTriplet`. See [document](https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md#vcpkgtriplet-triplet).

**Comment by jdx-john** [comment]
@JackBoosY those things are clear but does this mean if I am using a manifest file, I can only specify one triplet in each run? When specifying ports on the commandline I can do e.g.:
```
vcpkg install tinyxml:x86-windows tinyxml:x64-windows
```
But with manifest I'd have to run it twice:
```
vcpkg install --x-manifest-root=... --triplet=x86-windows
vcpkg install --x-manifest-root=... --triplet=x64-windows
```
This was raised, then closed, #7432, without explanation.

The other thing I was curious about was, I _cannot_ explicitly specify the manifest file _name_, only the _path_ to search for vcpkg.json? Would this be a reasonable feature-request

**Comment by JackBoosY** [comment]
cc @strega-nil-ms

**Comment by strega-nil-ms** [comment]
@jdx-john we do not support multiple triplets with manifest mode, since we don't think it's a useful feature. The only way one can do this is with "host triplets". Basically, your "target" triplet is used for libraries and the like, while your "host" triplet is used for binaries and tools.

Manifest mode is more for "one-tree-per-build", which means that you only _need_ one triplet for libraries (plus one for host tools).

**Comment by jdx-john** [comment]
Thanks @strega-nil I was confused about host/target triplets the other week but decided it wasn't something I needed to know :)

We can't be the only team who build the same solution twice with different configurations? e.g. we build Solution.sln twice, once for Release | x86 and once Release | x64. Or, many people must have a solution which has some x86 and some x64 projects - so it's likely you might need both x86 and x64 versions of the same libraries.

If it's possible to 'cheat' using host/target triplets to build against two triplets, could you give me an example how to do this?

**Comment by autoantwort** [comment]
We have the same problem. We set the install dir to `$(VcpkgRoot)installed$(VcpkgTriplet)` as a solution

**Comment by jdx-john** [comment]
Am I correct that when building with manifest, it wipes the target directory unlike when building explicit targets? 
My main use for manifest is to make it easier for us to pre-build all the dependencies, maybe I would be best explicitly building packages and put this in a script. I feel I'm fighting manifests a bit just using it as a list of packages to build.

**Comment by strega-nil-ms** [comment]
@jdx-john yeah, manifests is really _not_ about "just being a list of packages to install". If that's what you want, you're better off using a response file and classic mode. Manifest mode is about installing the packages you need for your current project and configuration, the same way other language package managers (like cargo) work.

The experience in CMake is likely better, since I was the one who implemented and I know CMake a _lot_ better. I think @autoantwort has the right idea tho, and I would like to figure out the best way to implement this in vcpkg proper by default.

**Comment by jdx-john** [comment]
OK I think we'll either use a simple script or look into response files (I hadn't seen those). We have a pretty giant solution so prefer our vcpkg dependencies are prebuilt in one place. Manifests sound like not the best fit here.

Thanks for clarifying.

**Comment by netcorefan1** [comment]
Discussion ended up nearly 2 years ago and I only saw people posting hacks. Not sure if this also applies to custom triplets, but for default ones (for example `x64-windows-static`) it looks like it has been implemented through expressions, but it doesn't seem to [work](https://github.com/microsoft/vcpkg/discussions/30641).

**Reply by dg0yt** [reply]
No hacks. Don't add confusion.
The triplet is a single consistent world. The user may choose which ports and features to install in this world. Platform expressions help to filter dependencies.
But only the triplet file defines the actual linkage. Triplet files may control linkage per port (within the range which is supported by the ports).
FTR triplet files can be provided in external directories via `--overlay-triplets=...`.

**Reply by netcorefan1** [reply]
Many thanks. Yes, I was confused and I wrongly believed that `(windows & x64 & static)` was the manifest way to do `vcpkg install x64-windows-static`.
@autoantwort posted a link with detailed instructions on how achieve per port customization.

**Comment by vipcxj** [comment]
Why make controlling linkage such a pain in the ass? Currently, the only way to control linkage is to rely on triplets, which control both platform and linkage, which I don't think is a very good design. Customizing your own linkage is a very common requirement, and coupling linkage with platform makes this common requirement a pain.

**Reply by dg0yt** [reply]
> Customizing your own linkage is a very common requirement.

I'm not convinced that it is *very common*. And when *customizing* is required, where is the problem with providing (or even generating) your own (extra or overlay) triplet file? In particular when it is so important to be done in a controlled and tested way.

**Reply by Codinablack** [reply]
Its for sure very common. Otherwise you wouldn't be finding so many issues centered around this same idea, not just on this github, but also on stackexchange, reddit and other places its being asked.

You keep saying, triplet that, and port this, but none of it makes sense from a novice user's perspective, and I'm not even that novice and yet I can't seem to get a straight answer on how I can alter my manifest file to make sure it gets vcpkg to install BOTH dynamic and static libraries... no instead, I keep finding links pointing me to the ridiculously shitty microsoft documentation on "custom triplets"... Why should I need to install an extra file or "triplet" or w/e the %$%# it is that is needed, onto the end users machine, just to get both static and dynamic libraries and offer both as configuration options when compiling for my project?? At this point, seems like the best option is to just not support both builds and only choose one, which is absolutely lame that it would have to come down to this.....

**Reply by Osyotr** [reply]
@Codinablack if you want to use static libraries, configure with static triplet. If you want use dynamic libraries, configure with dynamic triplet. You can't have both at the same time and this is not a vcpkg limitation.

**Reply by Codinablack** [reply]
That's funny because I found many instances where people give detailed instructions on how to use both static and dynamic for configurations... even seen somewhere they were giving specific GCC command line instructions to compile a project with mixed libs of static and dynamic, all I want is to use static for one configuration and dynamic for another

**Reply by Codinablack** [reply]
I don't understand why I can't have different libs linked based on configuration, nor why linux hates static linking so much

**Reply by Osyotr** [reply]
> all I want is to use static for one configuration and dynamic for another

Perhaps you may clarify things by providing sample project, but otherwise I don't see any reason why switching triplets based on configuration doesn't work for you.
